test/matrix.cc

```diff
@@ -198,64 +200,113 @@ TEST(Matrix, test_matrix) {
+    if (matrix.distances()[i] < kMaxCost) {
       ++found;
     }
   }
   EXPECT_EQ(found, 15) << " not the number of results as expected";
+  request.clear_matrix();
 
   CostMatrix cost_matrix_abort_target;
-  results =
-      cost_matrix_abort_target.SourceToTarget(*request.mutable_options()->mutable_sources(),
-                                              *request.mutable_options()->mutable_targets(), reader,
-                                              mode_costing, sif::TravelMode::kDrive, 50000.0);
+  cost_matrix_abort_target.SourceToTarget(request, reader, mode_costing, sif::TravelMode::kDrive,
+                                          50000.0);
 
+  matrix = request.matrix();
   found = 0;
-  for (uint32_t i = 0; i < results.size(); ++i) {
-    if (results[i].dist < kMaxCost) {
+  for (uint32_t i = 0; i < matrix.times().size(); ++i) {
+    if (matrix.distances()[i] < kMaxCost) {
       ++found;
     }
   }
   EXPECT_EQ(found, 10) << " not the number of results as expected";
+  request.clear_matrix();
 
   TimeDistanceMatrix timedist_matrix;
-  results = timedist_matrix.SourceToTarget(*request.mutable_options()->mutable_sources(),
-                                           *request.mutable_options()->mutable_targets(), reader,
-                                           mode_costing, sif::TravelMode::kDrive, 400000.0);
-  for (uint32_t i = 0; i < results.size(); ++i) {
-    EXPECT_NEAR(results[i].dist, matrix_answers[i].dist, kThreshold)
-        << "result " + std::to_string(i) + "'s distance is not equal to" +
-               " the expected value for TimeDistMatrix";
-
-    EXPECT_NEAR(results[i].time, matrix_answers[i].time, kThreshold)
-        << "result " + std::to_string(i) +
-               "'s time is not equal to the expected value for TimeDistMatrix";
+  timedist_matrix.SourceToTarget(request, reader, mode_costing, sif::TravelMode::kDrive, 400000.0);
+
+  matrix = request.matrix();
+  for (uint32_t i = 0; i < matrix.times().size(); ++i) {
+    EXPECT_NEAR(matrix.distances()[i], matrix_answers[i][1], kThreshold)
+        << "result " + std::to_string(i) + "'s distance is not equal" +
+               " to expected value for TDMatrix";
+
+    EXPECT_NEAR(matrix.times()[i], matrix_answers[i][0], kThreshold)
+        << "result " + std::to_string(i) + "'s time is not equal" + " to expected value for TDMatrix";
   }
 }
 
-TEST(Matrix, test_timedistancematrix_results_sequence) {
+TEST(Matrix, test_timedistancematrix_forward) {
+  // Input request is the same as `test_request`, but without the last target
+  const auto test_request_more_sources = R"({
+    "sources":[
+      {"lat":52.106337,"lon":5.101728},
+      {"lat":52.111276,"lon":5.089717},
+      {"lat":52.103105,"lon":5.081005}
+    ],
+    "targets":[
+      {"lat":52.106126,"lon":5.101497},
+      {"lat":52.100469,"lon":5.087099},
+      {"lat":52.103105,"lon":5.081005},
+      {"lat":52.094273,"lon":5.075254}
+    ],
+    "costing":"auto"
+  })";
```

Comment:
added a test for #4153

Reply:
The bug in #4153 only occurred with the bicycle profile, the auto profile worked fine. Does that matter for your test/fix?

Reply:
Jep that's fine. Over the API we decide internally which matrix algo to use. In your case that was TDMatrix, for auto it's always CostMatrix. This test uses TDMatrix deliberately. Good point though!